app/components/emoji_keyboard.js

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import React from 'react';
+import {
+    View,
+    Text,
+    StyleSheet,
+    FlatList,
+    TouchableOpacity
+} from 'react-native';
+import {KeyboardRegistry} from 'react-native-keyboard-input';
+import {store} from 'app/mattermost';
+import {CategoryNames, EmojiIndicesByCategory, Emojis} from 'app/utils/emojis';
+import Emoji from 'app/components/emoji/index';
+import PropTypes from 'prop-types';
+
+function buildCategory(category) {
+    return {
+        key: category,
+        data: EmojiIndicesByCategory.get(category).map(
+                (index) => Emojis[index].aliases[0])
+    };
+}
+const emojiCategories = CategoryNames.map(buildCategory);
+
+const EMOJI_SIZE = 32;
+const EMOJI_PADDING = 4;
+
+const styles = StyleSheet.create({
+    category: {
+        flex: 1,
+        alignItems: 'center',
+        justifyContent: 'center',
+        flexWrap: 'wrap',
+        alignContent: 'flex-end'
+    },
+    categoriesToolbar: {
+        flex: 0,
+        flexDirection: 'row'
+    }
+});
+
+class EmojiKeyboard extends React.PureComponent {
+    static propTypes = {
+        containerBackgroundColor: PropTypes.string.isRequired
+    }
+
+    static emptyLayout = () => ({layout: {}})
+    static emojiRange = [...(new Array(emojiCategories.length)).keys()]
+
+    static getCategoryNameByIndex = (index) => {
+        const nonNegativeIndex = Math.max(0, index);
+        const bounded = Math.min(emojiCategories.length - 1, nonNegativeIndex);
+        return emojiCategories[bounded].key;
+    }
+
+    state = {
+        layout: EmojiKeyboard.emojiRange.map(EmojiKeyboard.emptyLayout),
+        currentCategoryName: EmojiKeyboard.getCategoryNameByIndex(0)
+    }
+
+    onEmojiSelected = (emojiName) => {
+        KeyboardRegistry.onItemSelected('EmojiKeyboard', {emojiName});
+    }
+
+    scrollToCategoryByIndex = (index) => {
+        if (index === emojiCategories.length - 1) {
+            this.categoryListRef.scrollToEnd();
+        } else {
+            this.categoryListRef.scrollToIndex({index});
+        }
+    }
+
+    handleCategoryListRef = (ref) => {
+        this.categoryListRef = ref;
+    }
+
+    renderEmoji = (item, index) => (
+        <TouchableOpacity
+            key={index}
```

Comment:
can we set the key to other than the `index` just to avoid re-rendering if the index change